src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs

```diff
@@ -1004,14 +1122,731 @@ static void Main()
+                Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "F").WithArguments("Program.F<T>(T[])").WithLocation(6, 9),
+                // (7,12): error CS0103: The name 'Unknown' does not exist in the current context
+                //         F([Unknown]);
+                Diagnostic(ErrorCode.ERR_NameNotInContext, "Unknown").WithArguments("Unknown").WithLocation(7, 12),
+                // (8,9): error CS0411: The type arguments for method 'Program.F<T>(T[])' cannot be inferred from the usage. Try specifying the type arguments explicitly.
+                //         F([Main()]);
+                Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "F").WithArguments("Program.F<T>(T[])").WithLocation(8, 9));
+        }
+
+        [Fact]
+        public void TypeInference_32()
+        {
+            string source = """
+                delegate void D();
+                class Program
+                {
+                    static T[] F<T>(T[] x) => x;
+                    static void Main()
+                    {
+                        var x = F([null, Main]);
+                        x.Report(includeType: true);
+                        var y = F([Main, (D)Main]);
+                        y.Report(includeType: true);
+                    }
+                }
+                """;
+            CompileAndVerify(new[] { source, s_collectionExtensions }, expectedOutput: "System.Action[][null, System.Action], D[][D, D], ");
+        }
+
+        [Fact]
+        public void TypeInference_33()
+        {
+            string source = """
+                delegate byte D();
+                class Program
+                {
+                    static T[] F<T>(T[] x) => x;
+                    static void Main()
+                    {
+                        var x = F([null, () => 1]);
+                        x.Report(includeType: true);
+                        var y = F([() => 2, (D)(() => 3)]);
+                        y.Report(includeType: true);
+                    }
+                }
+                """;
+            CompileAndVerify(new[] { source, s_collectionExtensions }, expectedOutput: "System.Func<System.Int32>[][null, System.Func`1[System.Int32]], D[][D, D], ");
+        }
+
+        [Fact]
+        public void TypeInference_34()
+        {
+            string source = """
+                using System;
+                using System.Collections.Generic;
+                class Program
+                {
+                    static List<Func<T>> F1<T>(List<Func<T>> x) => x;
+                    static string F2() => null;
+                    static void Main()
+                    {
+                        var x = F1([F2]);
+                        var y = F1([null, () => 1]);
+                        var z = F1([() => default, F2]);
+                    }
+                }
+                """;
+            var comp = CreateCompilation(source);
+            // PROTOTYPE: Should inference succeed?
```

Comment:
It looks like `MethodTypeInferrer.MakeOutputTypeInferences()` should walk into collection literals, similar to the behavior for tuples.

Reply:
This is precisely what I was hoping to discover when asking for these tests, so I agree 😄.